### PR TITLE
Simplify 'buf generate' and fix insertion points

### LINKED ISSUE #600: panic in buf generate with misconfigured plugin
I added a new plugin `go-json_out` from https://github.com/mitchellh/protoc-gen-go-json and got some panics running `buf generate` until I got the actual error `Failure: could not find protoc plugin for name go-json_out` on the third try. That's when I realized that I needed to remove the `_out` suffix, after which the generate worked perfectly. Not a big deal, but probably worth fixing so that the error gets emitted correctly on the first run. Maybe also accepting or at least suggesting removing `_out` if detected?

### buf.gen.yaml

```
version: v1
managed:
  enabled: true
  go_package_prefix:
    default: go.nozzle.io
plugins:
  - name: go
    out: .
    opt: paths=source_relative
  - name: go-grpc
    out: .
    opt:
      - paths=source_relative
      - require_unimplemented_servers=false
  - name: go-json_out
    out: .
    opt:
      - paths=source_relative
      - enums_as_ints=false
      - emit_defaults=false
      - orig_name=false
      - allow_unknown=false
```

### Run 1

```
panic: send on closed channel

goroutine 110 [running]:
github.com/bufbuild/buf/private/buf/bufgen.semaphore.Acquire(0xc000d8f3e0, {0x1a99d68, 0xc001d37a00})
	github.com/bufbuild/buf/private/buf/bufgen/generator.go:584 +0x7a
github.com/bufbuild/buf/private/buf/bufgen.(*generator).generateConcurrently.func1()
	github.com/bufbuild/buf/private/buf/bufgen/generator.go:227 +0xa6
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x67
created by golang.org/x/sync/errgroup.(*Group).Go
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x92
panic: send on closed channel

goroutine 118 [running]:
github.com/bufbuild/buf/private/buf/bufgen.semaphore.Acquire(0xc000d8f3e0, {0x1a99d68, 0xc001d37a00})
	github.com/bufbuild/buf/private/buf/bufgen/generator.go:584 +0x7a
github.com/bufbuild/buf/private/buf/bufgen.(*generator).generateConcurrently.func1()
	github.com/bufbuild/buf/private/buf/bufgen/generator.go:227 +0xa6
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x67
created by golang.org/x/sync/errgroup.(*Group).Go
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x92
```

### Run 2

```
panic: send on closed channel

goroutine 93 [running]:
github.com/bufbuild/buf/private/buf/bufgen.semaphore.Acquire(0xc001b621e0, {0x1a99d68, 0xc002b7d640})
	github.com/bufbuild/buf/private/buf/bufgen/generator.go:584 +0x7a
github.com/bufbuild/buf/private/buf/bufgen.(*generator).generateConcurrently.func1()
	github.com/bufbuild/buf/private/buf/bufgen/generator.go:227 +0xa6
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x67
created by golang.org/x/sync/errgroup.(*Group).Go
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x92
panic: send on closed channel

goroutine 116 [running]:
github.com/bufbuild/buf/private/buf/bufgen.semaphore.Acquire(0xc001b621e0, {0x1a99d68, 0xc002b7d640})
	github.com/bufbuild/buf/private/buf/bufgen/generator.go:584 +0x7a
github.com/bufbuild/buf/private/buf/bufgen.(*generator).generateConcurrently.func1()
	github.com/bufbuild/buf/private/buf/bufgen/generator.go:227 +0xa6
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x67
created by golang.org/x/sync/errgroup.(*Group).Go
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x92
panic: send on closed channel

goroutine 124 [running]:
github.com/bufbuild/buf/private/buf/bufgen.semaphore.Acquire(0xc001b621e0, {0x1a99d68, 0xc002b7d640})
	github.com/bufbuild/buf/private/buf/bufgen/generator.go:584 +0x7a
github.com/bufbuild/buf/private/buf/bufgen.(*generator).generateConcurrently.func1()
	github.com/bufbuild/buf/private/buf/bufgen/generator.go:227 +0xa6
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x67
created by golang.org/x/sync/errgroup.(*Group).Go
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x92
panic: send on closed channel

goroutine 97 [running]:
github.com/bufbuild/buf/private/buf/bufgen.semaphore.Acquire(0xc001b621e0, {0x1a99d68, 0xc002b7d640})
	github.com/bufbuild/buf/private/buf/bufgen/generator.go:584 +0x7a
github.com/bufbuild/buf/private/buf/bufgen.(*generator).generateConcurrently.func1()
	github.com/bufbuild/buf/private/buf/bufgen/generator.go:227 +0xa6
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x67
created by golang.org/x/sync/errgroup.(*Group).Go
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x92
```

### Run 3

```
Failure: could not find protoc plugin for name go-json_out
```

### LINKED ISSUE #605: Insertion point output paths are not respected
Now that we're merging insertion points in-memory (for both local and remote generation), we aren't actually merging insertion points correctly. For `protoc`, insertion points actually work by referencing the output filepath of the to-be-generated file.

We have a pair of `protoc-gen-insertion-point-{receiver,writer}` defined that we can use for simple tests. Suppose I write the following `buf.gen.yaml`:

```yaml
version: v1
plugins:
  - name: insertion-point-receiver
    out: foo
  - name: insertion-point-writer
    out: bar
```

If I run `buf generate` with this configuration, I'll notice that the `insertion-point-writer` actually mutates the result of `insertion-point-receiver` and the result is deposited in the `foo` directory (even though `insertion-point-writer` was configured for the `bar` directory).

Plus, we actually print out a lot of unnecessary warnings around duplicate files since we aren't actually respecting insertion points [here](https://github.com/bufbuild/buf/blob/8d17256aee94cb59a52bdc4af03f06630e323d5c/private/pkg/app/appproto/response_writer.go#L75).

```
$ buf generate proto
Warning: Duplicate generated file name "test.txt". ...
...

$ tree 
.
├── bar
├── foo
│   └── test.txt
├── proto
│   └── ...
└── buf.gen.yaml
```

If I re-run the same command with `protoc`, you'll notice that it fails:

```sh
$ protoc -I proto proto/buf/alpha/module/v1alpha1/module.proto --insertion-point-receiver_out=foo --insertion-point-writer_out=bar
test.txt: Tried to insert into file that doesn't exist.
```

If I configure the `insertion-point-writer` to write to the same directory, it succeeds (as expected):

```
$ protoc -I proto proto/buf/alpha/module/v1alpha1/module.proto --insertion-point-receiver_out=foo --insertion-point-writer_out=foo
```

In short, we need to revisit [MergeInsertionPoints](https://github.com/bufbuild/buf/blob/ccc65fa5c97ce2b5209c15d28dc231779dee8eb2/private/bufpkg/bufplugin/bufplugin.go#L280) and verify that we are merging insertion points accurately with respect to the user's configured `out` for each local and/or remote plugin.

### PR DESCRIPTION
**Pull request by amckinney**
Fixes #600 and fixes #605

This revisits the `buf generate` command so that we consolidate all of our generation logic in `appproto` and `appprotoos`. With this, `buf protoc`, `buf generate`, and our server-side remote generation all use the same code paths.

This also fixes how we handle insertion points so that our behavior is aligned with `protoc`: we only permit insertion points to act upon other plugins configured to the same `_out` directory. For example,

```sh
$ protoc --insertion-point-receiver_out=gen/proto/insertion --insertion-point-writer_out=gen/proto/insertion
```

The `insertion-point-writer` will only have access to the files deposited in `gen/proto/insertion`, which is correctly configured above. Note that this was not previously the case.

Aside from the unit tests included here, this was tested against a variety of different cases, including combinations of `remote` and `local` plugins (including insertion points on either side).

### CODE REVIEW

File: private/buf/bufgen/generator.go [hunk 1]
```diff
@@ -173,239 +157,148 @@ func (g *generator) generateConcurrently(
...
+	//    - name: insertion-point-writer
+	//      out: gen/proto
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	if err := thread.Parallelize(
+		ctx,
+		jobs,
+		thread.ParallelizeWithCancel(cancel),
+		thread.ParallelizeWithFastFail(), // We want error messages to be concise here.
+	); err != nil {
```
[the remark this answers is not in the document]
Comment (amckinney) [reply private/buf/bufgen/generator.go:203]: These are interesting points, but I'd argue that the tradeoff of reinventing our own parallelization mechanism for this particular place in our codebase isn't worth it. One of our biggest priorities is consistency - `thread.Parallelize` might be susceptible to some of the issues you're describing, but we shouldn't go against the grain until we have a compelling reason to do so. By leaning into the `thread` package, we centralize all of our concurrency requirements in single place, where it's far more dependable and predictable.

In short, I'd prefer to remain as consistent, simple, and readable as possible until we hit a point where inventing something different is actually necessary. I could very well be wrong and that we'll need to refactor this later, but I think it's important that we optimize for consistency and don't prematurely optimize something that might not be an issue in practice.

[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply private/buf/bufgen/generator.go:203]: I think that's reasonable, but I worry that we have little observability into knowing whether this becomes a problem. Do we wait until users complain that remote plugins are undesireable because the take a long time? My biggest concern is making sure that the remote plugin requests are sent first, if we can do that and keep using `thread.Parallelize` it would be great. Maybe we should set a higher concurrency threshold for this execution?

[the remark this answers is not in the document]
Comment (bufdev) [reply private/buf/bufgen/generator.go:203]: We're optimizing for correct code over optimized code here - the primary purpose of this PR is to fix some outstanding bugs. We've QA'ed this, if we have an issue we'll revisit.

**Review by bufdev — APPROVED** [review]
